pytorch_lightning/trainer/connectors/accelerator_connector.py

```diff
@@ -73,959 +71,762 @@
+        elif self._accelerator_flag == "tpu":
+            if self._parallel_devices and len(self._parallel_devices) > 1:
+                self._strategy_flag = "tpu_spawn"
+            else:
+                # TODO lazy initialized device, then here could be self._strategy_flag = "single_tpu_device"
+                self._strategy_flag = SingleTPUStrategy(device=self._parallel_devices[0])
+        elif _HOROVOD_AVAILABLE and ("OMPI_COMM_WORLD_RANK" in os.environ or "HOROVOD_RANK" in os.environ):
+            self._strategy_flag = "horovod"
+        else:
+            if self._num_nodes_flag > 1:
+                self._strategy_flag = "ddp"
+            elif len(self._parallel_devices) <= 1:
+                device = (
+                    device_parser.determine_root_gpu_device(self._parallel_devices)
+                    if self._accelerator_flag == "gpu"
+                    else "cpu"
+                )
+                # TODO lazy initialized device, then here could be self._strategy_flag = "single_device"
+                self._strategy_flag = SingleDeviceStrategy(device=device)
+            elif len(self._parallel_devices) > 1:
+                self._strategy_flag = "ddp_spawn"
+            else:
+                self._strategy_flag = "ddp"
 
-    @property
-    def _is_sharded_training_type(self) -> bool:
-        return isinstance(self._strategy, (DDPShardedStrategy, DDPSpawnShardedStrategy))
+    def _strategy_check_and_fallbacks(self):
+        # current logic, fallback only apply to user pass in str config not object config
+        _strategy_flag = "" if isinstance(self._strategy_flag, Strategy) else self._strategy_flag
 
-    @property
-    def _is_fully_sharded_training_type(self) -> bool:
-        return isinstance(self._strategy, DDPFullyShardedStrategy)
+        if _strategy_flag == "ddp_cpu":
+            if _TPU_AVAILABLE:
+                raise MisconfigurationException(
+                    "`accelerator='ddp_cpu'` is not supported on TPU machines. "
+                    "Learn more: https://github.com/PyTorchLightning/pytorch-lightning/issues/7810"
+                )
+            if self._device_flag == 1 and self._num_nodes_flag > 1:
+                _strategy_flag = "ddp"
+            else:
+                _strategy_flag = "ddp_spawn"
+            if self._accelerator_flag == "gpu":
+                rank_zero_warn(
+                    "You requested one or more GPUs, but set `accelerator='ddp_cpu'`. Training will not use GPUs."
+                )
+        if _strategy_flag in ("ddp_spawn", "ddp_spawn_find_unused_parameters_false") and (
+            TorchElasticEnvironment.detect() or KubeflowEnvironment.detect() or self._is_slurm_managing_tasks()
+        ):
+            _strategy_flag = "ddp"
+        if _strategy_flag in ("dp", "ddp2") and self._accelerator_flag == "cpu":
+            rank_zero_warn(f"{_strategy_flag!r} is not supported on CPUs, hence setting `strategy='ddp'`.")
```

Comment:
why is this not a misconfiguration error? this should fail as fast as possible given how different DP and DDP2 are compared to DDP

Reply:
Yeah totally agree, this is current behavior, added to the follow up items 4.2 in #11449